src/widgets/scrollbar.rs

```diff
@@ -446,198 +449,177 @@ impl<'a> Scrollbar<'a> {
         }
     }
 
-    fn get_track_area(&self, area: Rect) -> Rect {
-        // Decrease track area if a begin arrow is present
-        let area = if self.begin_symbol.is_some() {
-            if self.is_vertical() {
-                // For vertical scrollbar, reduce the height by one
-                Rect::new(
-                    area.x,
-                    area.y + 1,
-                    area.width,
-                    area.height.saturating_sub(1),
-                )
-            } else {
-                // For horizontal scrollbar, reduce the width by one
-                Rect::new(
-                    area.x + 1,
-                    area.y,
-                    area.width.saturating_sub(1),
-                    area.height,
-                )
-            }
-        } else {
-            area
-        };
-        // Further decrease scrollbar area if an end arrow is present
-        if self.end_symbol.is_some() {
-            if self.is_vertical() {
-                // For vertical scrollbar, reduce the height by one
-                Rect::new(area.x, area.y, area.width, area.height.saturating_sub(1))
-            } else {
-                // For horizontal scrollbar, reduce the width by one
-                Rect::new(area.x, area.y, area.width.saturating_sub(1), area.height)
+    fn get_viewport_len(&self, area: Rect) -> u16 {
+        match self.orientation {
+            ScrollbarOrientation::VerticalRight | ScrollbarOrientation::VerticalLeft => area.height,
+            ScrollbarOrientation::HorizontalBottom | ScrollbarOrientation::HorizontalTop => {
+                area.width
             }
-        } else {
-            area
-        }
-    }
-
-    fn should_not_render(&self, track_start: u16, track_end: u16, content_length: usize) -> bool {
-        if track_end - track_start == 0 || content_length == 0 {
-            return true;
         }
-        false
     }
 
-    fn get_track_start_end(&self, area: Rect) -> (u16, u16, u16) {
+    /// Returns information about scrollbar track
+    ///
+    /// For ScrollbarOrientation::VerticalRight
+    ///
+    /// ```plain
+    ///                   ┌───────── track_axis
+    ///                   v
+    ///   ┌───────────────┐
+    ///   │               ║
+    ///   │               █
+    ///   │               █
+    ///   │               ║
+    ///   └───────────────┘
+    /// ```
+    ///
+    /// For ScrollbarOrientation::HorizontalBottom
+    ///
+    /// ```plain
+    ///   ┌───────────────┐
+    ///   │               │
+    ///   │               │
+    ///   │               │
+    ///   └═══███═════════┘<──────── track_axis
+    /// ```
+    fn get_track_axis(&self, area: Rect) -> u16 {
```

Comment:
To avoids introducing an unnecessary extra concept that exists only for the purpose of rendering (and all the associated commentary), this would be much simpler:

```rust
fn get_scollbar_area(&self, area: Rect) -> Rect {
  match self.orientation {
    ScrollbarOrientation::VerticalLeft => Rect { width: 1, ..area },
    ScrollbarOrientation::VerticalRight => Rect { x: area.right().saturating_sub(1), width: 1, ..area },
    ScrollbarOrientation::HorizontalTop => Rect { height: 1, ..area },
    ScrollbarOrientation::HorizontalBottom => Rect { y: area.bottom().saturating_sub(1), height: 1, ..area },
  }
}
```

Then this just gets used as:
```rust
for x in area.left()..area.right() {
  for y in area.top()..area..bottom() {
    let (symbol, style) = bar.next();
    buf.set_string(x, y, symbol, style);
  }
}
```
or an equivalent functional approach:
```rust
for (position, (symbol, style)) in area.rows().flatmap(Rect::columns).zip(bar) {
  buf.set_string(position.x, position.y, symbol, style);
}
```